cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp

```diff
@@ -0,0 +1,305 @@
+INSTANTIATE_TYPE_MAP(std::int8_t);
+INSTANTIATE_TYPE_MAP(std::int16_t);
+INSTANTIATE_TYPE_MAP(std::int32_t);
+INSTANTIATE_TYPE_MAP(std::int64_t);
+
+template <typename TestType>
+class select_flagged_test : public te::policy_fixture {
+public:
+    using Float = std::tuple_element_t<0, TestType>;
+    using Flag = std::tuple_element_t<1, TestType>;
+    using Integer = std::uint32_t;
+
+    auto allocate_arrays(Integer elem_count) {
+        auto& q = this->get_queue();
+        auto val = ndarray<Float, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+        auto mask = ndarray<Flag, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+
+        return std::make_tuple(val, mask);
+    }
+
+    auto allocate_vector_of_arrays(std::int64_t vector_count, Integer elem_count) {
+        auto& q = this->get_queue();
+        std::vector<ndarray<Float, 1>> val_vec(vector_count);
+        std::vector<ndarray<Flag, 1>> mask_vec(vector_count);
+        for (std::int64_t i = 0; i < vector_count; i++) {
+            val_vec[i] = ndarray<Float, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+            mask_vec[i] = ndarray<Flag, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+        }
+
+        return std::make_tuple(val_vec, mask_vec);
+    }
+
+    auto init_vector_of_arrays(std::vector<ndarray<Float, 1>>& val_vec,
+                               std::vector<ndarray<Flag, 1>>& mask_vec,
+                               Float a,
+                               Float b,
+                               std::int64_t seed = 777) {
+        const std::int64_t vector_count = val_vec.size();
+        for (std::int64_t i = 0; i < vector_count; i++) {
+            fill_uniform(val_vec[i], mask_vec[i], a, b, seed);
+        }
+
+        return std::make_tuple(val_vec, mask_vec);
+    }
+
+    void fill_uniform(ndarray<Float, 1>& val,
+                      ndarray<Flag, 1>& mask,
+                      Float a,
+                      Float b,
+                      std::int64_t seed = 777) {
+        Integer elem_count = de::integral_cast<Integer>(val.get_count());
+        std::mt19937 rng(seed);
+        std::uniform_real_distribution<Float> distr(a, b);
+
+        Float pivot = distr(rng);
+        // move generation to device when rng is available there
+        Float* val_ptr = detail::host_allocator<Float>().allocate(val.get_count());
+        Flag* mask_ptr = detail::host_allocator<Flag>().allocate(val.get_count());
+        for (Integer el = 0; el < elem_count; el++) {
+            val_ptr[el] = distr(rng);
+            mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
+        }
+        val.assign(this->get_queue(), val_ptr, val.get_count()).wait_and_throw();
+        mask.assign(this->get_queue(), mask_ptr, val.get_count()).wait_and_throw();
+
+        detail::host_allocator<Float>().deallocate(val_ptr, val.get_count());
+        detail::host_allocator<Flag>().deallocate(mask_ptr, val.get_count());
```

Comment:
It sounds like ndarray need to be created on host and next copied to device. All you need is `to_device`.
```c++
auto generate_uniform(std::int64_t count, Float a, Float b, int seed) const {
    std::mt19937 rng(seed);
    std::uniform_real_distribution<Float> distr(a, b);

    const auto val = ndarray<Float, 1>::empty(count);
    const auto mask = ndarray<Flag, 1>::empty(count); 

    Float* val_ptr = val.get_mutable_data();
    Flag* mask_ptr = mask.get_mutable_data();

    for (std::int64_t el = 0; el < count; el++) {
        val_ptr[el] = distr(rng);
        mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
    } 
    
    auto q = this->get_queue();
    return std::make_tuple(val.to_device(q), mask.to_device(q));
}
```

Reply:
In general, I see a lot of code copy-pasted in perf and functional tests. I think we could have common header, where base fuxture for these tests is defined. It's not problem of this particular primitive, but also sort and reduction to some extent. 

We shouldn't address this right now, but we need to keep in mind and plan refactoring iteration.

Reply:
added method to device, rewrote this code snippet accordingly, but keep "fill_uniform" naming for the case when generation can be done as separate step, for example in perf tests, where it is used for filling each element of vector of ndarrays.